File: leetcode/3612-process-string-with-special-operations-i/3612-process-string-with-special-operations-i.java
public class Solution {
    public String processStr(String s) {
        StringBuilder result = new StringBuilder();
        
        for (int i = 0; i < s.length(); i++) {
            char ch = s.charAt(i);
            
            if (Character.isLowerCase(ch)) {
                // Lowercase letters: Append to the result
                result.append(ch);
            } else if (ch == '*') {
                // Asterisk: Backspace operation
                if (result.length() > 0) {
                    result.deleteCharAt(result.length() - 1);
                }
            } else if (ch == '#') {
                // Hash: Duplicate the entire current result string
                result.append(result.toString());
            } else if (ch == '%') {
                // Percent: Reverse the entire current result string
                result.reverse();
            }
        }
        
        return result.toString();
    }
}

File: 3612-process-string-with-special-operations-i/3612-process-string-with-special-operations-i.ts
function processStr(s: string): string {
    const result: string[] = [];

    for (const ch of s) {
        if (ch >= 'a' && ch <= 'z') {
            result.push(ch);
        } else if (ch === '*') {
            if (result.length > 0) {
                result.pop();
            }
        } else if (ch === '#') {
            result.push(...result);
        } else if (ch === '%') {
            result.reverse();
        }
    }

    return result.join('');
}